### PR TITLE
feature: blockContentHex filter

### PR DESCRIPTION
**Pull request by szuecs**
feature: `blockContentHex("deadbeef", "00a0")` filter

### CODE REVIEW

**Comment by szuecs** [comment]
WIP, because I want to at least have more test coverage in lexer to test hex strings, there, too.

**Comment by szuecs** [comment]
```
# proxy
]% ./bin/skipper -inline-routes='r: * -> blockContent(`\x00`) -> "http://127.0.0.1:9001"'                                                                             feature/block-hex
[APP]INFO[0000] Expose metrics in codahale format
[APP]INFO[0000] enable swarm: false
[APP]INFO[0000] Replacing tee filter specification
[APP]INFO[0000] Replacing teenf filter specification
[APP]INFO[0000] Replacing lua filter specification
[APP]INFO[0000] support listener on :9911
[APP]INFO[0000] Dataclients are updated once, first load complete
[APP]INFO[0000] proxy listener on :9090
[APP]INFO[0000] TLS settings not found, defaulting to HTTP
[APP]INFO[0000] route settings, reset, route: r: * -> blockContent("00") -> "http://127.0.0.1:9001"
[APP]INFO[0000] route settings received
[APP]INFO[0000] route settings applied
[APP]ERRO[0002] error while proxying after 1.434815ms, route r with backend network http://127.0.0.1:9001, status code 400: dialing failed false: readfrom tcp 127.0.0.1:33898->127.0.0.1:9001: blocked string match found in body, remote host: ::1, request: "POST / HTTP/1.1", host: localhost:9090, user agent: "curl/7.49.0"
::1 - - [04/May/2023:21:41:58 +0200] "POST / HTTP/1.1" 400 12 "-" "curl/7.49.0" 1 localhost:9090 - -

# backend
% ./bin/skipper -inline-routes='* -> status(200) -> inlineContent("foo") -> <shunt>' -address :9001                                                                  feature/block-hex
[APP]INFO[0000] Expose metrics in codahale format
[APP]INFO[0000] enable swarm: false
[APP]INFO[0000] Replacing tee filter specification
[APP]INFO[0000] Replacing teenf filter specification
[APP]INFO[0000] Replacing lua filter specification
[APP]INFO[0000] support listener on :9911
[APP]INFO[0000] Dataclients are updated once, first load complete
[APP]INFO[0000] proxy listener on :9001
[APP]INFO[0000] TLS settings not found, defaulting to HTTP
[APP]INFO[0000] route settings, reset, route: : * -> status(200) -> inlineContent("foo") -> <shunt>
[APP]INFO[0000] route settings received
[APP]ERRO[0000] Failed to start supportListener on :9911: listen tcp :9911: bind: address already in use
[APP]INFO[0000] route settings applied
[APP]ERRO[2616] error while discarding remainder request body: unexpected EOF.
127.0.0.1 - - [04/May/2023:21:41:58 +0200] "POST / HTTP/1.1" 499 3 "-" "curl/7.49.0" 0 127.0.0.1:9001 - -

# client
% echo "\x00\x00\x00x00" | curl http://localhost:9090 -vvv --data-binary @-                                                                  feature/skipper-read-and-write-timeout-filters
* Rebuilt URL to: http://localhost:9090/
*   Trying ::1...
* Connected to localhost (::1) port 9090 (#0)
> POST / HTTP/1.1
> Host: localhost:9090
> User-Agent: curl/7.49.0
> Accept: */*
> Content-Length: 7
> Content-Type: application/x-www-form-urlencoded
>
* upload completely sent off: 7 out of 7 bytes
< HTTP/1.1 400 Bad Request
< Content-Length: 12
< Content-Type: text/plain; charset=utf-8
< Server: Skipper
< X-Content-Type-Options: nosniff
< Date: Thu, 04 May 2023 19:41:58 GMT
<
Bad Request
```

File: eskip/lexer.go [hunk 1]
```diff
@@ -125,13 +126,47 @@ func scanVoid(code string, p charPredicate) string {
...
+					secondBinValue = false
+					i, err := strconv.ParseInt(string(c), 16, 8)
+					if err != nil {
+						binary = false
+						code = code[1:]
+						continue
+					}
+					bin[1] = byte(i)
+				}
+			}
```
Comment (AlexanderYastrebov) [inline eskip/lexer.go:168]: I am not sure it is actually working. I do not see we store parsed binary bytes into returned result `b`  anywehere which is confirmed by this test that fails:
```go
func TestStringHex(t *testing.T) {
	for _, tc := range []struct {
		filter, expected string
	}{
		{`foo("\x00")`, "\x00"},
		{`foo("a\x00")`, "a\x00"},
		{`foo("\x00b")`, "\x00b"},
		{`foo("a\x00b")`, "a\x00b"},

		{`foo("\x000")`, "\x000"},
	} {
		t.Run(tc.filter, func(t *testing.T) {
			ff, err := ParseFilters(tc.filter)
			if err != nil {
				t.Fatal(err)
			}

			s := ff[0].Args[0]

			if s != tc.expected {
				t.Errorf("wrong string value, expected %q: got: %q", tc.expected, s)
			}
		})
	}
}
```

We also need to adjust eskip serialization logic to escape binary values somehow I guess.

Comment (szuecs) [reply eskip/lexer.go:168]: how do you think the filter blockContent would be able to block the tested content?

File: docs/reference/filters.md [hunk 2]
```diff
@@ -976,6 +976,23 @@ Example:
 * -> blockContent("Malicious Content") -> "http://example.com";
 ```
 
+### blockHexContent
```
Comment (AlexanderYastrebov) [inline docs/reference/filters.md:979]: Lets maybe use `Hex` suffix, i.e. `blockContentHex` similar to `repeatContentHex` and `wrapContentHex`

Comment (szuecs) [reply docs/reference/filters.md:979]: done

**Comment by szuecs** [comment]
:+1:

**Pull request merged**